Fix/rounding changes

jcompagni10 opened the pull request:
Change rounding to favor maker instead of taker. This is necessary to eliminate a large class of rounding-related errors and is a more optimal UX for market makers. 
As part of this change, we also enforce a trueLimitPrice to ensure taker LimitOrders are being executed at the intended price.

jcompagni10 committed:
reverse rounding behavior; favor maker

jcompagni10 committed:
Enforce true limitprice on LimitOrders

jcompagni10 committed:
only use ErrLimitPriceNotSatisfied for consistency

jcompagni10 committed:
fix ibc tests

jcompagni10 committed:
Merge remote-tracking branch 'origin/main' into fix/rounding_changes

jcompagni10 committed:
multihop test fixes

jcompagni10 committed:
lint fixes

pr0n00gler commented:
Please fix linter issues

jcompagni10 committed:
Add cap to withdraw amounts + tests

jcompagni10 committed:
format

jcompagni10 committed:
Merge remote-tracking branch 'origin/main' into fix/rounding_changes

NeverHappened commented on x/dex/keeper/liquidity.go:
```diff
@@ -130,15 +115,69 @@ func (k Keeper) SaveLiquidity(sdkCtx sdk.Context, liquidityI types.Liquidity) {
-	pctDiff := priceDiffFromExpected.Quo(bookPrice)
-
-	return pctDiff.GT(maxTrueTakerSpread)
+func (k Keeper) TakerLimitOrderSwap(
```
some comments describing takerLimitOrderSwap and MakerLimitOrder swap and why there is such distinction would be nice

NeverHappened commented on x/dex/keeper/multihop_swap.go:
```diff
@@ -34,7 +34,7 @@ func (k Keeper) HopsToRouteData(
 		price, found := k.GetCurrPrice(ctx, tradePairID)
 		if !found {
-			return routeArr, types.ErrInsufficientLiquidity
+			return routeArr, types.ErrLimitPriceNotSatisfied
```
maybe different error? it's kinda more about missing liquidity pools at all than limit price not satisfied. (yeah i know probably a rare occurence, but still)

jcompagni10 replied:
Decided to simplify things by having a single error here. From a user's perspective, if there's no liquidity or all of the liquidity is above their limitPrice, it's functionally the same.

NeverHappened commented on x/dex/keeper/liquidity.go:
```diff
@@ -43,20 +42,6 @@ func (k Keeper) Swap(
-		// being swapped, it is possible that the next swap could yield a "fair" price.
-		// Nonetheless, once the remainingTakerDenom gets small enough to start causing unfair swaps
-		// it is much simpler to just abort.
-		if inAmount.IsZero() || isUnfairTruePrice(params.MaxTrueTakerSpread, inAmount, outAmount, liq) {
```
now that we removed this check - i see you've added the check for truePrice < exitLimitPrice.
Do we have any tests for cases where this condition is not satisfied?

jcompagni10 replied:
https://github.com/neutron-org/neutron/blob/fix%2Frounding_changes/x/dex/keeper/integration_placelimitorder_test.go#L471
https://github.com/neutron-org/neutron/blob/fix%2Frounding_changes/x/dex/keeper/integration_placelimitorder_test.go#L304

NeverHappened commented on x/dex/keeper/core.go:
```diff
@@ -331,49 +331,38 @@ func (k Keeper) PlaceLimitOrderCore(
+	// This is ok because tokenOut is provided to the constructor of PairID above
+	takerTradePairID := pairID.MustTradePairIDFromMaker(tokenOut)
+	var limitPrice math_utils.PrecDec
+	limitPrice, err = types.CalcPrice(tickIndexInToOut)
```
a little out of scope, but how about renaming tickIndexInToOut -> limitTickIndexInToOut? if I understand it correctly,  it's acting as a tick limit, not as a tick of an actual swap

NeverHappened reviewed (commented):
nice!
I also suggest adding some tests from here https://github.com/neutron-org/neutron/pull/443/files
And we can add a test that checks multihop swap works on cheap to expensive routes

jcompagni10 committed:
fix withdraw arithmetic

jcompagni10 committed:
formatting

jcompagni10 committed:
add additional dust/rounding test

jcompagni10 committed:
Add comments

jcompagni10 committed:
Remove ErrExitLimitPriceHit and just use ErrLimitPriceNotSatisfied

there is no need for 2 different errors that mean the same thing

jcompagni10 committed:
typos, etc.

jcompagni10 committed:
remove unnceccessary comment

jcompagni10 committed:
switch to UnixSecs for ExpirationKey

This avoids a bug caused by unixMs for JIT 0 time wrapping into a huge number and
not being indexed correctly. Also saves a bit of case by shortening the key.
Since the millisecond precision is way overkill

pr0n00gler commented:
Please fix linter issues, otherwise lgtm

jcompagni10 committed:
lint / format

jcompagni10 committed:
Merge branch 'feat/sdk-50' into fix/rounding_changes

NeverHappened commented on x/dex/keeper/liquidity.go:
```diff
@@ -130,15 +115,73 @@ func (k Keeper) SaveLiquidity(sdkCtx sdk.Context, liquidityI types.Liquidity) {
+
+	truePrice := math_utils.NewPrecDecFromInt(totalOutCoin.Amount).QuoInt(totalInCoin.Amount)
+
+	if truePrice.LT(limitPrice) {
```
Is this correct, or should be the other way around?

jcompagni10 commented:
https://github.com/neutron-org/neutron-tests/actions/runs/9276692068

jcompagni10 committed:
Merge remote-tracking branch 'origin/feat/sdk-50' into fix/rounding_changes

jcompagni10 committed:
Merge remote-tracking branch 'origin/feat/sdk-50' into fix/rounding_changes

swelf19 commented on x/dex/types/keys.go:
```diff
@@ -110,8 +110,13 @@ func LimitOrderTrancheUserAddressPrefix(address string) []byte {
-	str := utils.Uint64ToSortableString(unixMs)
+	var unixSecs uint64
+	// If timestamp is 0 use that instead of returning long negative number for unix time
+	if !timestamp.IsZero() {
```
1) i still not sure i understand what is it for
2) by changing timebytes representation, we change `LimitOrderExpirationKey`, do we need migration for existing orders?

jcompagni10 committed:
Merge remote-tracking branch 'origin/feat/sdk-50' into fix/rounding_changes

jcompagni10 committed:
add migration

jcompagni10 committed:
disallow maker or taker limit orders that will result in < 1 token of output

This is a nice ux improvement for taker orders that would otherwise get a
less informative ErrLimitPriceNotSatisfied.

This is a critical fix for maker orders that would otherwise not be withdrawable
since a withdrawable amount between 0 and 1 will be rounded to 0 throwing a ErrWithdrawEmptyLimitOrder

jcompagni10 committed:
Don't use maker leg for filled limit orders.

This is a waste of gas and will almost certainly hit a ErrTradeTooSmall which is
undesirable

jcompagni10 committed:
Merge remote-tracking branch 'origin/feat/sdk-50' into fix/rounding_changes
